Fix PETSc vector serialization.

As reported on the mailing list:
```
/home/user/Downloads/git-files/dealii/include/deal.II/lac/petsc_vector_base.h:1324:55: error: no matching constructor for initialization of 'boost::serialization::array_wrapper<const double>'
    boost::serialization::array_wrapper<const double> wrapper(
                                                      ^
/media/storage/local_opt/boost/include/boost/serialization/array_wrapper.hpp:46:5: note: candidate constructor not viable: no known conversion from 'const PetscScalar *' (aka 'const complex<double> *') to 'const double *' for 1st argument
    array_wrapper(T * t, std::size_t s) :
    ^
/media/storage/local_opt/boost/include/boost/serialization/array_wrapper.hpp:41:5: note: candidate constructor not viable: requires single argument 'rhs', but 2 arguments were provided
    array_wrapper(const array_wrapper & rhs) :
    ^
In file included from /home/user/Downloads/git-files/dealii/build/source/numerics/unity_0.cc:10:
In file included from /home/user/Downloads/git-files/dealii/source/numerics/dof_output_operator.cc:23:
In file included from /home/user/Downloads/git-files/dealii/include/deal.II/lac/petsc_block_vector.h:27:
In file included from /home/user/Downloads/git-files/dealii/include/deal.II/lac/petsc_vector.h:29:
/home/user/Downloads/git-files/dealii/include/deal.II/lac/petsc_vector_base.h:1363:49: error: no matching constructor for initialization of 'boost::serialization::array_wrapper<double>'
    boost::serialization::array_wrapper<double> velocity_wrapper(
                                                ^
/media/storage/local_opt/boost/include/boost/serialization/array_wrapper.hpp:46:5: note: candidate constructor not viable: no known conversion from 'PetscScalar *' (aka 'complex<double> *') to 'double *' for 1st argument
    array_wrapper(T * t, std::size_t s) :
    ^
/media/storage/local_opt/boost/include/boost/serialization/array_wrapper.hpp:41:5: note: candidate constructor not viable: requires single argument 'rhs', but 2 arguments were provided
    array_wrapper(const array_wrapper & rhs) :
    ^
```

And @masterleinad also pointed out the solution:

> It looks like
> 
>   boost::serialization::array_wrapper<const double> wrapper(
> 
> should be replaced with
> 
>   boost::serialization::array_wrapper<const PetscScalar> wrapper(
> 
> and similarly in VectorBase::load. 
> 

I think this is probably a consequence of #15945.

--- a/include/deal.II/lac/petsc_vector_base.h
+++ b/include/deal.II/lac/petsc_vector_base.h
@@ -1321,7 +1321,7 @@ namespace PETScWrappers
     int                ierr  = VecGetArrayRead(*this, &array);
     AssertThrow(ierr == 0, ExcPETScError(ierr));
 
-    boost::serialization::array_wrapper<const double> wrapper(
+    boost::serialization::array_wrapper<const PetscScalar> wrapper(
       array, locally_owned_size());
     ar &wrapper;
 
@@ -1360,9 +1360,9 @@ namespace PETScWrappers
     int          ierr  = VecGetArray(petsc_vector(), &array);
     AssertThrow(ierr == 0, ExcPETScError(ierr));
 
-    boost::serialization::array_wrapper<double> velocity_wrapper(
+    boost::serialization::array_wrapper<PetscScalar> wrapper(
       array, locally_owned_size());
-    ar &velocity_wrapper;
+    ar &wrapper;
 
     ierr = VecRestoreArray(petsc_vector(), &array);
     AssertThrow(ierr == 0, ExcPETScError(ierr));

